### ISSUE TITLE
How to use efficient_risk properly?

### ISSUE THREAD

**Issue by PyDataBlog** [robertmartin8/PyPortfolioOpt#156]
**What are you trying to do?**
I am simply trying to get the best weights for a specified risk level using the `efficient_risk` method. I keep getting an optimisation error. What could I be doing wrong? Is this a bug? Here's a reproducible code:

```py
import yfinance as yf
import matplotlib.pyplot as plt
import pandas as pd
from pypfopt import EfficientFrontier, risk_models, expected_returns  # version '1.2.4'


tickers = ["MSFT", "AMZN", "KO", "MA", "COST", 
           "LUV", "XOM", "PFE", "JPM", "UNH", 
           "ACN", "DIS", "GILD", "F", "TSLA"]

ohlc = yf.download(tickers, period="max")
prices = ohlc["Adj Close"]

mu = expected_returns.capm_return(prices)
S = risk_models.CovarianceShrinkage(prices).ledoit_wolf()

# Constrained efficient risk
ef = EfficientFrontier(mu, S)
ef.efficient_risk(target_volatility=0.10, market_neutral=False)

```
I get this error:

```py
NoneType: None
---------------------------------------------------------------------------
OptimizationError                         Traceback (most recent call last)
<ipython-input-72-182e530cb4a3> in <module>
----> 1 ef.efficient_risk(target_volatility=0.10, market_neutral=False)

~/miniconda3/envs/ef/lib/python3.8/site-packages/pypfopt/efficient_frontier.py in efficient_risk(self, target_volatility, market_neutral)
    286             self._constraints.append(cp.sum(self._w) == 1)
    287 
--> 288         return self._solve_cvxpy_opt_problem()
    289 
    290     def efficient_return(self, target_return, market_neutral=False):

~/miniconda3/envs/ef/lib/python3.8/site-packages/pypfopt/base_optimizer.py in _solve_cvxpy_opt_problem(self)
    219             raise exceptions.OptimizationError
    220         if opt.status != "optimal":
--> 221             raise exceptions.OptimizationError
    222         self.weights = self._w.value.round(16) + 0.0  # +0.0 removes signed zero
    223         return self._make_output_weights()

OptimizationError: Please check your objectives/constraints or use a different solver.
```

**Comment by robertmartin8** [comment]
Could you show me what your covariance matrix looks like? Sometimes this
issue occurs when the optimise cannot find a portfolio that matches your
target risk.

**Comment by PyDataBlog** [comment]
@robertmartin8 
Here's the covariance matrix of shape (15, 15):

```py
Symbols      MSFT      AMZN        KO        MA      COST       LUV       XOM       PFE       JPM       UNH       ACN       DIS      GILD         F      TSLA
Symbols                                                                                                                                                      
MSFT     0.069056  0.042779  0.019963  0.043520  0.023712  0.030498  0.027588  0.022824  0.036848  0.032976  0.036933  0.030201  0.024566  0.029575  0.045804
AMZN     0.042779  0.092488  0.012986  0.040297  0.021393  0.024888  0.020299  0.018331  0.026678  0.026036  0.029370  0.025335  0.025473  0.024233  0.052567
KO       0.019963  0.012986  0.033346  0.022923  0.014904  0.022016  0.020315  0.016407  0.022902  0.019253  0.020792  0.020185  0.013965  0.021579  0.019744
MA       0.043520  0.040297  0.022923  0.068805  0.021779  0.038129  0.033866  0.024506  0.042559  0.034542  0.039302  0.036245  0.025276  0.038995  0.046837
COST     0.023712  0.021393  0.014904  0.021779  0.038274  0.017762  0.015572  0.015760  0.020394  0.020198  0.020761  0.018530  0.017841  0.016566  0.019075
LUV      0.030498  0.024888  0.022016  0.038129  0.017762  0.109106  0.029880  0.021538  0.045268  0.029650  0.031533  0.035819  0.021811  0.041592  0.037370
XOM      0.027588  0.020299  0.020315  0.033866  0.015572  0.029880  0.056788  0.019659  0.039623  0.028598  0.028381  0.029590  0.019476  0.036434  0.032639
PFE      0.022824  0.018331  0.016407  0.024506  0.015760  0.021538  0.019659  0.041303  0.025962  0.025351  0.021112  0.019108  0.025509  0.020367  0.018457
JPM      0.036848  0.026678  0.022902  0.042559  0.020394  0.045268  0.039623  0.025962  0.070742  0.035892  0.037460  0.037842  0.025479  0.047389  0.037566
UNH      0.032976  0.026036  0.019253  0.034542  0.020198  0.029650  0.028598  0.025351  0.035892  0.068577  0.029751  0.026406  0.021193  0.031719  0.032471
ACN      0.036933  0.029370  0.020792  0.039302  0.020761  0.031533  0.028381  0.021112  0.037460  0.029751  0.053367  0.030485  0.022232  0.031209  0.036015
DIS      0.030201  0.025335  0.020185  0.036245  0.018530  0.035819  0.029590  0.019108  0.037842  0.026406  0.030485  0.057924  0.020935  0.036775  0.038566
GILD     0.024566  0.025473  0.013965  0.025276  0.017841  0.021811  0.019476  0.025509  0.025479  0.021193  0.022232  0.020935  0.082633  0.020557  0.031939
F        0.029575  0.024233  0.021579  0.038995  0.016566  0.041592  0.036434  0.020367  0.047389  0.031719  0.031209  0.036775  0.020557  0.088159  0.048277
TSLA     0.045804  0.052567  0.019744  0.046837  0.019075  0.037370  0.032639  0.018457  0.037566  0.032471  0.036015  0.038566  0.031939  0.048277  0.292582
```

**Comment by robertmartin8** [comment]
@PyDataBLog

Yeah so if you look at the diagonal of your covariance matrix, the smallest I can see is about 0.03. The diagonal entries are variances, so the volatilities are given by the square root of that, which is about 0.17. Hence there is no single-asset portfolio that can match a target of 0.10. It's true that a portfolio of assets can be less risky if you have anticorrelated assets, but that isn't the case here. 

TL;DR: Use a higher target_risk, probably around 0.18-0.20 should work.

**Comment by PyDataBlog** [comment]
@robertmartin8 that's very interesting. A neat check could be making sure that users receive a warning if they request for an unrealistic target with a check like this: 

```py
import numpy as np

assert target_volatility >= np.diag(np.sqrt(S)).min() 
```

That way only workable tasks are passed. What do you think?

**Comment by robertmartin8** [comment]
@PyDataBlog 

You're right: that would be a useful feature. However, it wouldn't be as simple as your snippet because the portfolio volatility can be lower than the minimum asset volatility (e.g if you have an anticorrelated asset), so one actually needs to check the eigenvalues. I will have a think about how best to do this.

Cheers,
Robert

**Comment by hkelley44** [comment]
I am getting the same error as @PyDataBlog except I know that the target risk level is achievable. Here the code snippet:
![image](https://user-images.githubusercontent.com/31108706/95523181-d2163000-0982-11eb-8d71-67db6982071c.png)

It's strange because it fails on particular target risk during the iterations but not when I use that target risk outside of the loop: 
![image](https://user-images.githubusercontent.com/31108706/95523260-07bb1900-0983-11eb-8fe1-ca367c2e3b22.png)

Here is the posterior covariance matrix: 
![image](https://user-images.githubusercontent.com/31108706/95523290-186b8f00-0983-11eb-820f-7701863c7b4c.png)

Any help would be greatly appreciated! Thanks
